File: Data Structures & Algorithms/balanced-binary-tree/submission-2.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {boolean}
     */
    isBalanced(root) {
        const dfs = (root) => {
            if (!root) return 0;

            let left = dfs(root.left)
            let right = dfs(root.right)

            if (Math.abs(left - right) <= 1) {
                return 1 + Math.max(left, right)
            } else {
                return;
            }
        }

        return dfs(root) >= 0
    }
}

File: Data Structures & Algorithms/depth-of-binary-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        def dfs(root, count):
            if root is None:
                return count

            left_depth = dfs(root.left, count + 1)
            right_depth = dfs(root.right, count + 1)

            return max(left_depth, right_depth)
        return dfs(root, 0)
